File: frontend/web/src/features/masters/configs/unitConversion.ts
import { ArrowLeftRight } from 'lucide-react';
import type { MasterEntityConfig } from '../engine/types';

export const unitConversionConfig: MasterEntityConfig = {
  key: 'unitConversion',
  label: 'Unit Conversion',
  labelPlural: 'Unit Conversions',
  description: 'Conversion factors between two Units of Measure (e.g. 1 Box = 10 Strip).',
  icon: ArrowLeftRight,
  section: 'Units & Tax',
  fields: [
    { key: 'fromUomId', label: 'From Unit', type: 'reference', referenceEntityKey: 'unitOfMeasure', required: true },
    { key: 'toUomId', label: 'To Unit', type: 'reference', referenceEntityKey: 'unitOfMeasure', required: true },
    { key: 'conversionFactor', label: 'Conversion Factor', type: 'decimal', required: true, min: 0, step: 0.000001, helpText: '1 "From Unit" = this many "To Unit". From and To cannot be the same unit.' },
  ],
  getDisplayLabel: (record, resolveReference) =>
    `${resolveReference('unitOfMeasure', record.fromUomId as string)} → ${resolveReference('unitOfMeasure', record.toUomId as string)}`,
  validateForm: (values) =>
    values.fromUomId && values.toUomId && values.fromUomId === values.toUomId
      ? 'From Unit and To Unit cannot be the same.'
      : undefined,
};

File: backend/src/Modules/Masters/HMS.Modules.Masters/Domain/UnitConversion.cs
using HMS.Shared.Kernel;

namespace HMS.Modules.Masters.Domain;

/// <summary>
/// Conversion factor between two <see cref="UnitOfMeasure"/> records (docs/03_Masters_ERD,
/// "Units &amp; Tax"), e.g. 1 Box = 10 Strip. Has no natural code/name of its own — displayed
/// as "{FromUom} → {ToUom}" by callers, mirroring the frontend's getDisplayLabel.
/// </summary>
internal class UnitConversion : Entity
{
    public Guid FromUomId { get; private set; }
    public Guid ToUomId { get; private set; }
    public decimal ConversionFactor { get; private set; }

    public bool IsActive { get; private set; } = true;

    private UnitConversion()
    {
    }

    private UnitConversion(Guid id, Guid fromUomId, Guid toUomId, decimal conversionFactor, bool isActive, Guid? createdBy)
        : base(id, createdBy)
    {
        FromUomId = fromUomId;
        ToUomId = toUomId;
        ConversionFactor = conversionFactor;
        IsActive = isActive;
    }

    public static UnitConversion Create(Guid fromUomId, Guid toUomId, decimal conversionFactor, bool isActive, Guid? createdBy)
    {
        if (fromUomId == toUomId)
        {
            throw new ArgumentException("From Unit and To Unit cannot be the same.");
        }

        if (conversionFactor <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(conversionFactor), "Conversion factor must be greater than 0.");
        }

        return new UnitConversion(Guid.CreateVersion7(), fromUomId, toUomId, conversionFactor, isActive, createdBy);
    }

    public void Update(decimal conversionFactor, bool isActive, Guid? updatedBy)
    {
        if (conversionFactor <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(conversionFactor), "Conversion factor must be greater than 0.");
        }

        ConversionFactor = conversionFactor;
        IsActive = isActive;
        MarkUpdated(updatedBy);
    }
}
